### ISSUE TITLE
[BUG:] When switching to new versions of LLamaSharp 0.16.0, there was a slowdown

### ISSUE THREAD

**Issue by aropb** [SciSharp/LLamaSharp#921]
### Description

After upgrading from KM 0.61 and LLamaSharp 0.13 to the latest versions, the processing ImportDocumentAsync/AskAsync speed has significantly decreased, almost 2 times.
All settings and equipment are the same. I see that the GPU is now less loaded, previously CUDA was loaded more than 80%.

Import:
![Screenshot_1](https://github.com/user-attachments/assets/d069ca43-b104-4850-8020-cfa7826efbf0)

Ask:
![image](https://github.com/user-attachments/assets/7bb512f2-9695-46ee-87b3-64a77f9eec1a)

How do I get back to my previous speed?
Thanks.

### Reproduction Steps

-

### Environment & Configuration

Operating system: Windows 11
.NET runtime version: 8.0.8
LLamaSharp version: 0.16.0
KernelMemory: 0.71...
CUDA version (if you are using cuda backend): 12
CPU & GPU device: CPU or GPU, Intel Core Ultra 9 & RTX 3090

**Comment by aropb** [comment]
Launched Llama

![image](https://github.com/user-attachments/assets/556d8a0d-f964-4731-b8af-cfcf6ad74829)

**Comment by aropb** [comment]
apparently there was a similar problem
https://github.com/SciSharp/LLamaSharp/issues/601

**Comment by aropb** [comment]
GPU utilization is not more than 30%.

This my test code:

            ModelParams = new(Program.AppSettings.ModelPath)
            {
                Embeddings = false,
                ContextSize = 8192,
                Seed = 1337,
                MainGpu = 0,
                GpuLayerCount = 100,
                SplitMode = GPUSplitMode.Layer,
                BatchSize = 1024,
                UBatchSize = 1024
            };

            Weights = LLamaWeights.LoadFromFile(ModelParams);
            StatelessExecutor executor = new(Weights, ModelParams);
            
            InferenceParams inferenceParams = new()
            {
                SamplingPipeline = new DefaultSamplingPipeline()
                {
                    Temperature = 0.5f,
                    RepeatPenalty = 1.0f
                },
                MaxTokens = 2000,
                AntiPrompts = ["\n"]
            };

            StringBuilder sb = new(2048);

            await foreach (string text in executor.InferAsync(value, inferenceParams, cancellationToken))
            {
                sb.Append(text);
            }

            return sb.ToString();

**Comment by martindevans** [comment]
What model are you using, how many tokens are you processing at once, and how are you measuring GPU utilisation?

**Comment by aropb** [comment]
Model [ggml-model-Q4_K_M.gguf]
https://huggingface.co/ruslandev/llama-3-8b-gpt-4o-ru1.0-gguf/blob/main/ggml-model-Q4_K_M.gguf

In the text above, the value line is: come up with and write a poem about summer.

Measuring GPU utilisation: the graphs are higher and the most important thing is to increase the response time by 2 times, create embeddings

**Comment by aropb** [comment]
This problem only manifests itself on the GPU

**Comment by martindevans** [comment]
The default values in `DefaultSamplingPipeline` should be the same as the default sampling values in the llama.cpp `main` application. Typical P [is documented](https://github.com/ggerganov/llama.cpp/tree/master/examples/main#locally-typical-sampling) as having a default value of `1.0`. So it looks like that's a bug in the default values for `DefaultSamplingPipeline.`.

Would you like to submit a PR changing that default value? If so, and if you want to go put in some extra effort, checking _all_ of the default values against the readme I linked above would be a good idea :)

> And why does this parameter have such an effect?

All the parameter is doing is enabling a new sampling stage in llama.cpp (it's disabled entirely when the value is 1.0), so presumably it's just that expensive to calculate typical-p.

**Comment by zsogitbe** [comment]
These two parameters are too high (should decrease to 512 - this is the optimal value for speed and GPU memory use):
```
BatchSize = 2048,
UBatchSize = 2048
```

**Comment by aropb** [comment]
It is very easy for you to carefully read everything that I have written. I spent a lot of time on this.And there are already a lot of questions. The speed has increased, utilization is 60%, but still not 80% as it should be. I moved all the default settings as they should be. You should install them in the constructor for others. Bath size 512 I can't install because this size is not enough for me. And for some reason, it is checked for the equality of batchsize and ubatchsize.

**Comment by aropb** [comment]
I use KM. When creating embeddings, a value less than batchsize=2048 does not work, it throws an error. The problem is not even this, but this:

public LLamaEmbedder(LLamaWeights weights, IContextParams @params, ILogger? logger = null)
    {
        **if (@params.UBatchSize != @params.BatchSize)**
        {
            throw new ArgumentException("For non-causal models, batch size must be equal to ubatch size", "params");
        }


    }

**Comment by zsogitbe** [comment]
I am using batch size 512 with KM with different models having different vector size and I have never had a problem. 512 is the optimal value for speed and GPU memory use. Your problem will be somewhere else!

**Comment by aropb** [comment]
I'll check it again. Are you using 512/512?

**Comment by aropb** [comment]
The issue of GPU utilization is still relevant.

**Comment by aropb** [comment]
> I'll check it again. Are you using 512/512? ContextSize?

Input contains more tokens than configured batch size (Parameter 'batch')

**Comment by martindevans** [comment]
> @params.UBatchSize != @params.BatchSize
This is inherited from [here](https://github.com/ggerganov/llama.cpp/blob/master/examples/embedding/embedding.cpp#L88-L89).

Unfortunately the embedder can't split input into multiple batches at the moment, so if you are getting embeddings for huge chunks of text you will need an unusually large batch size. Again, this is inherited from that code linked above, which the new embedder is basically ported from.

> The issue of GPU utilization is still relevant.

Just a note about this, I think the measurements you're showing might be a little misleading. The typical-p thing gives us a clue - that was adding extra work on the CPU side (sampling is not done on the GPU). So that doesn't lower the actual peak utilisation, but it _does_ increase the amount of idle GPU time between inference (while sampling is running). Slightly different, it's just showing as lower utilisation because the sampling time in task manager is too low.

**Comment by aropb** [comment]
But I'm comparing Lama native and LLamaSharp graphics on the same machine, in the same configuration.

**Comment by aropb** [comment]
OK, but why does LLama native show more than 80%?

**Comment by zsogitbe** [comment]
> > @params.UBatchSize != @params.BatchSize
> > This is inherited from [here](https://github.com/ggerganov/llama.cpp/blob/master/examples/embedding/embedding.cpp#L88-L89).
> 
> Unfortunately the embedder can't split input into multiple batches at the moment, so if you are getting embeddings for huge chunks of text you will need an unusually large batch size. Again, this is inherited from that code linked above, which the new embedder is basically ported from.

It does not have any sense to use huge chunks of text for embeddings. The optimal size of embeddings is between 100-500 words depending on the task (document retrieval is on the lower side and summarization is on the higher side). So, maximum around 500 words. This should fit into a normal batch size of 512. If you decrease these numbers your GPU utilization will increase.

**Comment by aropb** [comment]
I use a chunk size of 1000 words. Experimentally, this is the best size for me.

The GPU loading problem manifests itself with normal output, not embedding.

**Comment by aropb** [comment]
Embedding comes first, then a break, then GenerateText:

![image](https://github.com/user-attachments/assets/682bdbe5-5e8e-48cb-829b-733fe5fb0f1a)

**Comment by martindevans** [comment]
Embedding and text generation aren't really comparable workloads. Embedding processes one chunk of text all at once (it's similar to prompt processing). Wheras text generation has to constantly go back and forth betwen GPU (generating logits) and CPU (sampling logits to produce tokens). You'd expect GPU load to be lower in that case.

**Comment by aropb** [comment]
It was just additional information. You persistently ignore the important things that I write :)
I am comparing the utilization when working with Lama native and LLamaSharp when doing the same job.

**Comment by aropb** [comment]
I'm unlikely to find a reason, because I don't have that deep knowledge of LM and llama native. I thought that my information would help the experts understand where the performance problem might be. Can you compare GPU utilization on your PC with lama native & lama sharp?

**Comment by martindevans** [comment]
I'm unlikely to be able to find time in September for anything more than issue triage and PR reviews unfortunately. I'll be happy to dig into perf stuff in October though (probably after doing a 0.17 update).

**Comment by aropb** [comment]
Thank you very much for your work! I am ready to participate in improving LLamaSharp as much as my knowledge allows.

**Comment by aropb** [comment]
We also need to bring everything to the use of SamplingPipeline:

StatelessExecutor.InferAsync()
...
var repeat_last_n = Math.Max(0, inferenceParams.RepeatLastTokensCount < 0 ? _weights.ContextSize : inferenceParams.RepeatLastTokensCount);
...
inferenceParams.RepeatLastTokensCount obsolete !!!

**Comment by aropb** [comment]
New Version 0.17.0:

![image](https://github.com/user-attachments/assets/d2a8e921-6174-4a09-a2a0-a36f0364362b)
